File: NeetCode150/arrays_and_hashing/2_valid_anagram.cc
// Time: O(n)
// Space: O(n)
class Solution {
    public:
        bool isAnagram(string s, string t) {
            // first check the lengths, if not equal return false;
            // then create a map from s including counts,
            // if any letter is not used in the end return false;
    
            if (s.size() != t.size()) return false;
    
            unordered_map<char, int> map;
    
            for (auto ch: s) {
                map[ch]++;
            }
    
            for (auto ch: t) {
                map[ch]--;
                if (map[ch] < 0) return false;
            }
    
            return true;
        }
    };

File: NeetcodeBlind75/arrays_and_hashing/2_valid_anagram.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        """
        Algorithm intuition:
            - check if size of s and t is different
                -> return False if different
            - if we only have english letters, we can declare a constant size array (len 26) and count
            - first we count for s
            - then subtract each ord of character occurenece in t
            - at any time if any character count is less than 0 we return False
            - we do one more iteration through s to check if any character count is greater than 0, and return False
            - if all conditions pass we return true
        Time: O(n) where n = max(len(s), len(t))
        Space: O(1)
        """

        if len(s) != len(t):
            return False

        counts = [0] * 26

        for ch in s:
            counts[ord(ch) - ord("a")] += 1

        for ch in t:
            counts[ord(ch) - ord("a")] -= 1

        for c in counts:
            if c != 0:
                return False

        return True
